README
=======================================================================
 INSTALL
=======================================================================

 1. cd build
 2. cmake .. -DBUILD_SHARED_LIBS=ON -DCMAKE_BUILD_TYPE=Release

 parameters
    - BUILD_SHARED_LIBS (ON|OFF)
    - CMAKE_BUILD_TYPE (Debug|Release)
    - CMAKE_INSTALL_PREFIX (/usr/local)

3. make
4. make test
5. make install


Or just simply,

perl ./scripts/install

=======================================================================
 USAGE
=======================================================================

#include <geohash.h>

/* Get Hash */
char *hash;
hash = GEOHASH_encode(latitude, longitude, length);
...
free(hash);


/* Get Area by Hash */
GEOHASH_area *area;
area = GEOHASH_decode("dqcw5");

/* You can get the range of both latitude and longitude

area->latitude.max;
area->latitude.min;
area->longitude.max;
area->longitude.min;

*/
GEOHASH_free_area(area);


/* Get Adjacent Hash by Origin's Hash and Direction*/
char *adjacent_hash;
/* You can choose direction from GEOHASH_NORTH, GEOHASH_SOUTH, GEOHASH_EAST, and GEOHASH_WEST */
adjacent_hash = GEOHASH_get_adjacent("dqcw5", GEOHASH_NORTH);
...
free(adjacent_hash);



/* Get Neighbors' Hash by Origin's Hash */
GEOHASH *neighbors;
neighbors = GEOHASH_get_neighbors("dqcw5");

/* You can get hashes of each direction's neighbor
neighbors->north;
neighbors->north_east;
neighbors->north_west;
neighbors->south;
neighbors->south_east;
neighbors->south_west;
neighbors->east;
neighbors->west;
*/
GEOHASH_free_neighbors(neighbors);


debian/changelog
geohash (1.0.1) stable; urgency=low

  * Initial packaging effort

 -- Ricky Cormier <evilrix@gmail.com>  Tue, 23 Jul 2013 00:00:00 +0100

geohash (1.0.0) stable; urgency=low

  * Initial version

 -- Lyo Kato <lyo.kato@gmail.com>  Tue, 23 Jul 2013 00:00:00 +0100


tests/geohash_test.c
#include <stdlib.h>
#include <stdio.h>

#include <CUnit/CUnit.h>
#include <CUnit/Console.h>
#include <CUnit/Basic.h>

#include "geohash/geohash.h"

void verify_hash(double lat, double lon, int len, const char* expected)
{
    char *hash;
    hash = GEOHASH_encode(lat, lon, len);
    CU_ASSERT_STRING_EQUAL(hash, expected);
    free(hash);
    /*CU_ASSERT_PTR_NULL();*/
}

void test_geohash_encode(void)
{
    verify_hash(       45.37,       -121.7,  6, "c216ne"        );
    verify_hash(  47.6062095, -122.3320708, 13, "c23nb62w20sth" );
    verify_hash(  35.6894875,  139.6917064, 13, "xn774c06kdtve" );
    verify_hash( -33.8671390,  151.2071140, 13, "r3gx2f9tt5sne" );
    verify_hash(  51.5001524,   -0.1262362, 13, "gcpuvpk44kprq" );
}

void verify_area(
        const char *hash,
        double lat_min, double lon_min,
        double lat_max, double lon_max)
{
    GEOHASH_area *area;
    area = GEOHASH_decode(hash);
    CU_ASSERT_DOUBLE_EQUAL(area->latitude.max,   lat_max, 0.001);
    CU_ASSERT_DOUBLE_EQUAL(area->latitude.min,   lat_min, 0.001);
    CU_ASSERT_DOUBLE_EQUAL(area->longitude.max, lon_max, 0.001);
    CU_ASSERT_DOUBLE_EQUAL(area->longitude.min, lon_min, 0.001);
    GEOHASH_free_area(area);
}

void test_geohash_decode(void)
{
    verify_area("c216ne", 45.3680419921875, -121.70654296875, 45.37353515625, -121.695556640625);
    verify_area("C216Ne", 45.3680419921875, -121.70654296875, 45.37353515625, -121.695556640625);
    verify_area("dqcw4", 39.0234375, -76.552734375, 39.0673828125, -76.5087890625);
    verify_area("DQCW4", 39.0234375, -76.552734375, 39.0673828125, -76.5087890625);
}

void verify_adjacent(const char *origin, GEOHASH_direction dir, const char *expected)
{
    char *hash;
    hash = GEOHASH_get_adjacent(origin, dir);
    CU_ASSERT_STRING_EQUAL(hash, expected);
    free(hash);
}

void test_geohash_adjacent(void)
{
    verify_adjacent("dqcjq", GEOHASH_NORTH, "dqcjw");
    verify_adjacent("dqcjq", GEOHASH_SOUTH, "dqcjn");
    verify_adjacent("dqcjq", GEOHASH_WEST,  "dqcjm");
    verify_adjacent("dqcjq", GEOHASH_EAST,  "dqcjr");
}

void verify_neighbors(
        const char *origin,
        const char *hash1,
        const char *hash2,
        const char *hash3,
        const char *hash4,
        const char *hash5,
        const char *hash6,
        const char *hash7,
        const char *hash8
        )
{
    GEOHASH_neighbors *neighbors;
    neighbors = GEOHASH_get_neighbors(origin);

    CU_ASSERT_STRING_EQUAL(neighbors->north,      hash1);
    CU_ASSERT_STRING_EQUAL(neighbors->south,      hash2);
    CU_ASSERT_STRING_EQUAL(neighbors->west,       hash3);
    CU_ASSERT_STRING_EQUAL(neighbors->east,       hash4);
    CU_ASSERT_STRING_EQUAL(neighbors->north_west, hash5);
    CU_ASSERT_STRING_EQUAL(neighbors->north_east, hash6);
    CU_ASSERT_STRING_EQUAL(neighbors->south_west, hash7);
    CU_ASSERT_STRING_EQUAL(neighbors->south_east, hash8);

    GEOHASH_free_neighbors(neighbors);
}

void test_geohash_neighbors(void)
{
    verify_neighbors("dqcw5", "dqcw7", "dqctg", "dqcw4", "dqcwh", "dqcw6", "dqcwk", "dqctf", "dqctu");
    verify_neighbors("xn774c", "xn774f", "xn774b", "xn7749", "xn7751", "xn774d", "xn7754", "xn7748", "xn7750");
    verify_neighbors("gcpuvpk", "gcpuvps", "gcpuvph", "gcpuvp7", "gcpuvpm", "gcpuvpe", "gcpuvpt", "gcpuvp5", "gcpuvpj");
    verify_neighbors("c23nb62w", "c23nb62x", "c23nb62t", "c23nb62q", "c23nb62y", "c23nb62r", "c23nb62z", "c23nb62m", "c23nb62v");
}

void test_geohash_verification(void)
{
    CU_ASSERT_EQUAL(GEOHASH_verify_hash("dqcw5"), 1);
    CU_ASSERT_EQUAL(GEOHASH_verify_hash("dqcw7"), 1);
    CU_ASSERT_EQUAL(GEOHASH_verify_hash("abcwd"), 0);
    CU_ASSERT_EQUAL(GEOHASH_verify_hash("dqcw5@"), 0);
}

int main( /* int argc, char **argv */ )
{
    CU_pSuite suite;

    CU_initialize_registry();

    suite = CU_add_suite("GeoHashTestSuite", NULL, NULL);

    CU_add_test( suite, "GeoHashDecodeTest",       test_geohash_decode       );
    CU_add_test( suite, "GeoHashEncodeTest",       test_geohash_encode       );
    CU_add_test( suite, "GeoHashAdjacentTest",     test_geohash_adjacent     );
    CU_add_test( suite, "GeoHashNeighborsTest",    test_geohash_neighbors    );
    CU_add_test( suite, "GeoHashVerificationTest", test_geohash_verification );

    CU_basic_set_mode(CU_BRM_VERBOSE);
    CU_basic_run_tests();

    CU_cleanup_registry();

    return 0;
}


src/main.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

#include <geohash/geohash.h>

void show_usage()
{
    fprintf(stderr, "usage: \n");
    fprintf(stderr, "    geohash 'hash'\n\n");
    fprintf(stderr, "example: \n");
    fprintf(stderr, "    geohash c216ne\n");
    fprintf(stderr, "    geohash c23nb62w20sth\n");
    fprintf(stderr, "    geohash xn774c06kdtve\n");
}

void show_area(const char *hash)
{
    GEOHASH_area *area;
    area = GEOHASH_decode(hash);
    printf("=========================================\n");
    printf("AREA\n");
    printf("-----------------------------------------\n");
    printf("MAX LATITUDE:   %f\n", area->latitude.max);
    printf("MIN LATITUDE:   %f\n", area->latitude.min);
    printf("MAX longitude: %f\n", area->longitude.max);
    printf("MIN longitude: %f\n", area->longitude.min);
    GEOHASH_free_area(area);
}

void show_neighbors(const char *hash)
{

    GEOHASH_neighbors *neighbors;
    neighbors = GEOHASH_get_neighbors(hash);
    printf("=========================================\n");
    printf("NEIGHBORS\n");
    printf("-----------------------------------------\n");
    printf("NORTH:        %s\n", neighbors->north);
    printf("SOUTH:        %s\n", neighbors->south);
    printf("WEST:         %s\n", neighbors->west);
    printf("EAST:         %s\n", neighbors->east);
    printf("NORTH_WEST:   %s\n", neighbors->north_west);
    printf("NORTH_EAST:   %s\n", neighbors->north_east);
    printf("SOUTH_WEST:   %s\n", neighbors->south_west);
    printf("SOUTH_EAST:   %s\n", neighbors->south_east);
    GEOHASH_free_neighbors(neighbors);
}

int main(int argc, char** argv)
{
    if (argc < 2) {
        fprintf(stderr, "no argument\n\n");
        show_usage();
        return EXIT_FAILURE;
    }

    if (!GEOHASH_verify_hash(argv[1])) {
        fprintf(stderr, "invalid hash: %s\n\n", argv[1]);
        show_usage();
        return EXIT_FAILURE;
    }

    show_area(argv[1]);
    show_neighbors(argv[1]);

    return EXIT_SUCCESS;
}



include/geohash.h
/*
The MIT License

Copyright (c) 2011 lyo.kato@gmail.com

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
*/

#ifndef _LIB_GEOHASH_H_
#define _LIB_GEOHASH_H_

#include <stdbool.h>

#if defined(__cplusplus)
extern "C" {
#endif

typedef enum {
  GEOHASH_NORTH = 0,
  GEOHASH_EAST,
  GEOHASH_WEST,
  GEOHASH_SOUTH
} GEOHASH_direction;

typedef struct {
    double max;
    double min;
} GEOHASH_range;

typedef struct {
    GEOHASH_range latitude;
    GEOHASH_range longtitude;
} GEOHASH_area;

typedef struct {
    char* north;
    char* east;
    char* west;
    char* south;
    char* north_east;
    char* south_east;
    char* north_west;
    char* south_west;
} GEOHASH_neighbors;

bool GEOHASH_verify_hash(const char *hash);
char* GEOHASH_encode(double latitude, double longtitude, unsigned int hash_length);
GEOHASH_area* GEOHASH_decode(const char* hash);
GEOHASH_neighbors* GEOHASH_get_neighbors(const char *hash);
void GEOHASH_free_neighbors(GEOHASH_neighbors *neighbors);
char* GEOHASH_get_adjacent(const char* hash, GEOHASH_direction dir);
void GEOHASH_free_area(GEOHASH_area *area);

#if defined(__cplusplus)
}
#endif

#endif


src/geohash.c
/*
The MIT License

Copyright (c) 2011 lyo.kato@gmail.com

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
*/

#include "geohash.h"
#include <ctype.h>
#include <string.h>
#include <stddef.h>
#include <stdlib.h>
#include <stdbool.h>
#include <assert.h>

#define MAX_HASH_LENGTH 22

#define REFINE_RANGE(range, bits, offset) \
    if (((bits) & (offset)) == (offset)) \
        (range)->min = ((range)->max + (range)->min) / 2.0; \
    else \
        (range)->max = ((range)->max + (range)->min) / 2.0;

#define SET_BIT(bits, mid, range, value, offset) \
    mid = ((range)->max + (range)->min) / 2.0; \
    if ((value) >= mid) { \
        (range)->min = mid; \
        (bits) |= (0x1 << (offset)); \
    } else { \
        (range)->max = mid; \
        (bits) |= (0x0 << (offset)); \
    }

static const char BASE32_ENCODE_TABLE[33] = "0123456789bcdefghjkmnpqrstuvwxyz";
static const char BASE32_DECODE_TABLE[44] = {
    /* 0 */   0, /* 1 */   1, /* 2 */   2, /* 3 */   3, /* 4 */   4,    
    /* 5 */   5, /* 6 */   6, /* 7 */   7, /* 8 */   8, /* 9 */   9,    
    /* : */  -1, /* ; */  -1, /* < */  -1, /* = */  -1, /* > */  -1, 
    /* ? */  -1, /* @ */  -1, /* A */  -1, /* B */  10, /* C */  11, 
    /* D */  12, /* E */  13, /* F */  14, /* G */  15, /* H */  16, 
    /* I */  -1, /* J */  17, /* K */  18, /* L */  -1, /* M */  19, 
    /* N */  20, /* O */  -1, /* P */  21, /* Q */  22, /* R */  23, 
    /* S */  24, /* T */  25, /* U */  26, /* V */  27, /* W */  28, 
    /* X */  29, /* Y */  30, /* Z */  31    
};

static const char NEIGHBORS_TABLE[8][33] = {
    "p0r21436x8zb9dcf5h7kjnmqesgutwvy", /* NORTH EVEN */
    "bc01fg45238967deuvhjyznpkmstqrwx", /* NORTH ODD  */
    "bc01fg45238967deuvhjyznpkmstqrwx", /* EAST EVEN  */
    "p0r21436x8zb9dcf5h7kjnmqesgutwvy", /* EAST ODD   */
    "238967debc01fg45kmstqrwxuvhjyznp", /* WEST EVEN  */
    "14365h7k9dcfesgujnmqp0r2twvyx8zb", /* WEST ODD   */
    "14365h7k9dcfesgujnmqp0r2twvyx8zb", /* SOUTH EVEN */
    "238967debc01fg45kmstqrwxuvhjyznp"  /* SOUTH ODD  */
};

static const char BORDERS_TABLE[8][9] = {
    "prxz",     /* NORTH EVEN */
    "bcfguvyz", /* NORTH ODD */
    "bcfguvyz", /* EAST  EVEN */
    "prxz",     /* EAST  ODD */
    "0145hjnp", /* WEST  EVEN */
    "028b",     /* WEST  ODD */
    "028b",     /* SOUTH EVEN */
    "0145hjnp"  /* SOUTH ODD */
};

bool
GEOHASH_verify_hash(const char *hash)
{
    const char *p;
    unsigned char c;
    p = hash;
    while (*p != '\0') {
        c = toupper(*p++);
        if (c < 0x30)
            return false;
        c -= 0x30;
        if (c > 43)
            return false;
        if (BASE32_DECODE_TABLE[c] == -1)
            return false;
    }
    return true;
}

GEOHASH_area* 
GEOHASH_decode(const char *hash)
{
    const char *p;
    unsigned char c;
    char bits;
    GEOHASH_area *area;
    GEOHASH_range *range1, *range2, *range_tmp;

    area = (GEOHASH_area *)malloc(sizeof(GEOHASH_area));
    if (area == NULL)
        return NULL;

    area->latitude.max   =   90; 
    area->latitude.min   =  -90; 
    area->longitude.max =  180; 
    area->longitude.min = -180;

    range1 = &area->longitude;
    range2 = &area->latitude;

    p = hash;

    while (*p != '\0') {

        c = toupper(*p++);
        if (c < 0x30) {
            free(area);
            return NULL;
        }
        c -= 0x30;
        if (c > 43) {
            free(area);
            return NULL;
        }
        bits = BASE32_DECODE_TABLE[c];
        if (bits == -1) {
            free(area);
            return NULL;
        }

        REFINE_RANGE(range1, bits, 0x10);
        REFINE_RANGE(range2, bits, 0x08);
        REFINE_RANGE(range1, bits, 0x04);
        REFINE_RANGE(range2, bits, 0x02);
        REFINE_RANGE(range1, bits, 0x01);

        range_tmp = range1;
        range1    = range2;
        range2    = range_tmp;
    }
    return area;
}

char* 
GEOHASH_encode(double lat, double lon, unsigned int len)
{
    int i;
    char *hash;
    unsigned char bits = 0;
    double mid;
    GEOHASH_range lat_range = {  90,  -90 };
    GEOHASH_range lon_range = { 180, -180 };

    double val1, val2, val_tmp;
    GEOHASH_range *range1, *range2, *range_tmp;

    assert(lat >= -90.0);
    assert(lat <= 90.0);
    assert(lon >= -180.0);
    assert(lon <= 180.0);
    assert(len <= MAX_HASH_LENGTH);

    hash = (char *)malloc(sizeof(char) * (len + 1));
    if (hash == NULL)
        return NULL;

    val1 = lon; range1 = &lon_range;
    val2 = lat; range2 = &lat_range;

    for (i=0; i < len; i++) {

        bits = 0;
        SET_BIT(bits, mid, range1, val1, 4);
        SET_BIT(bits, mid, range2, val2, 3);
        SET_BIT(bits, mid, range1, val1, 2);
        SET_BIT(bits, mid, range2, val2, 1);
        SET_BIT(bits, mid, range1, val1, 0);

        hash[i] = BASE32_ENCODE_TABLE[bits];
        
        val_tmp   = val1;
        val1      = val2;
        val2      = val_tmp;
        range_tmp = range1;
        range1    = range2;
        range2    = range_tmp;
    }

    hash[len] = '\0';
    return hash;
}

void 
GEOHASH_free_area(GEOHASH_area *area)
{
    free(area);
}

GEOHASH_neighbors*
GEOHASH_get_neighbors(const char* hash)
{
    GEOHASH_neighbors *neighbors;

    neighbors = (GEOHASH_neighbors*)malloc(sizeof(GEOHASH_neighbors));
    if (neighbors == NULL)
        return NULL;

    neighbors->north = GEOHASH_get_adjacent(hash, GEOHASH_NORTH);
    neighbors->east  = GEOHASH_get_adjacent(hash, GEOHASH_EAST);
    neighbors->west  = GEOHASH_get_adjacent(hash, GEOHASH_WEST);
    neighbors->south = GEOHASH_get_adjacent(hash, GEOHASH_SOUTH);

    neighbors->north_east = GEOHASH_get_adjacent(neighbors->north, GEOHASH_EAST);
    neighbors->north_west = GEOHASH_get_adjacent(neighbors->north, GEOHASH_WEST);
    neighbors->south_east = GEOHASH_get_adjacent(neighbors->south, GEOHASH_EAST);
    neighbors->south_west = GEOHASH_get_adjacent(neighbors->south, GEOHASH_WEST);

    return neighbors;
}

char*
GEOHASH_get_adjacent(const char* hash, GEOHASH_direction dir)
{
    int len, idx;
    const char *border_table, *neighbor_table;
    char *base, *refined_base, *ptr, last;

    assert(hash != NULL);

    len  = strlen(hash);
    last = tolower(hash[len - 1]);
    idx  = dir * 2 + (len % 2);

    border_table = BORDERS_TABLE[idx];

    base = (char *)malloc(sizeof(char) * (len + 1));
    if (base == NULL)
        return NULL;
    memset(base, '\0', sizeof(char) * (len + 1));

    strncpy(base, hash, len - 1);

    if (strchr(border_table, last) != NULL) {
        refined_base = GEOHASH_get_adjacent(base, dir);
        if (refined_base == NULL) {
            free(base);
            return NULL;
        }
        strncpy(base, refined_base, strlen(refined_base));
        free(refined_base);
    }

    neighbor_table = NEIGHBORS_TABLE[idx];

    ptr = strchr(neighbor_table, last);
    if (ptr == NULL) {
        free(base);
        return NULL;
    }
    idx = (int)(ptr - neighbor_table);
    len = strlen(base);
    base[len] = BASE32_ENCODE_TABLE[idx];
    return base;
}

void 
GEOHASH_free_neighbors(GEOHASH_neighbors *neighbors)
{
    free(neighbors->north);
    free(neighbors->east);
    free(neighbors->west);
    free(neighbors->south);
    free(neighbors->north_east);
    free(neighbors->south_east);
    free(neighbors->north_west);
    free(neighbors->south_west);
    free(neighbors);
}

